kubelet/datadog_checks/kubelet/kubelet.py

```diff
@@ -0,0 +1,456 @@
+            headers = {'Authorization': 'Bearer {}'.format(self.kubelet_conn_info['token'])}
+
+        return requests.get(url, timeout=timeout, verify=verify,
+                            cert=cert, headers=headers, params={'verbose': verbose})
+
+    def retrieve_pod_list(self):
+        return self.perform_kubelet_query(self.pod_list_url).json()
+
+    def retrieve_node_spec(self):
+        """
+        Retrieve node spec from kubelet.
+        """
+        node_spec = self.perform_kubelet_query(self.node_spec_url).json()
+        # TODO: report allocatable for cpu, mem, and pod capacity
+        # if we can get it locally or thru the DCA instead of the /nodes endpoint directly
+        return node_spec
+
+    def _report_node_metrics(self, instance_tags):
+        node_spec = self.retrieve_node_spec()
+        num_cores = node_spec.get('num_cores', 0)
+        memory_capacity = node_spec.get('memory_capacity', 0)
+
+        tags = instance_tags
+        self.gauge(self, self.NAMESPACE + '.cpu.capacity', float(num_cores), tags)
+        self.gauge(self, self.NAMESPACE + '.memory.capacity', float(memory_capacity), tags)
+
+    def _perform_kubelet_check(self, instance_tags):
+        """Runs local service checks"""
+        service_check_base = self.NAMESPACE + '.kubelet.check'
+        is_ok = True
+        url = self.kube_health_url
+
+        try:
+            req = self.perform_kubelet_query(url)
+            for line in req.iter_lines():
+                # avoid noise; this check is expected to fail since we override the container hostname
+                if line.find('hostname') != -1:
+                    continue
+
+                matches = re.match(r'\[(.)\]([^\s]+) (.*)?', line)
+                if not matches or len(matches.groups()) < 2:
+                    continue
+
+                service_check_name = service_check_base + '.' + matches.group(2)
+                status = matches.group(1)
+                if status == '+':
+                    self.service_check(service_check_name, AgentCheck.OK, tags=instance_tags)
+                else:
+                    self.service_check(service_check_name, AgentCheck.CRITICAL, tags=instance_tags)
+                    is_ok = False
+
+        except Exception as e:
+            self.log.warning('kubelet check %s failed: %s' % (url, str(e)))
+            self.service_check(service_check_base, AgentCheck.CRITICAL,
+                               message='Kubelet check %s failed: %s' % (url, str(e)), tags=instance_tags)
+        else:
+            if is_ok:
+                self.service_check(service_check_base, AgentCheck.OK, tags=instance_tags)
+            else:
+                self.service_check(service_check_base, AgentCheck.CRITICAL, tags=instance_tags)
+
+    def _report_pods_running(self, pods, instance_tags):
+        """
+        Reports the number of running pods on this node
+        tagged by service and creator.
+        """
+        for pod in pods['items']:
```

Comment:
IIRC this will fail if `pods == None`, we should check that `pods != None and 'items' in pods` first